### ISSUE TITLE
Stream interface?

### ISSUE THREAD

**Issue by brianshaler** [derhuerst/hifo#2]
While the way it does it now is correct, in the sense that only the top `size` items will appear in the result, I think it would be interesting to have a stream where items passed through are the ones that have been added, in any order.

``` js
dataStream()
.pipe(hifo(hifo.highest(), 10))
.pipe(res) // would likely result in >10 items, but that can be okay
```

This could be seen as "incorrect" because the first `size` items would be passed through, but would allow the results to continue in a stream pipeline and potentially get re-filtered down to `size` items upon presentation.

My use case is a leveldb+websockets city name autocomplete feature, where it may take a couple seconds for leveldb to finish streaming all `t*` cities, and I would prefer not to wait until the leveldb stream completes before I send the first result to the client. Currently, I have a primitive, domain-specific solution in place, where I filter results by a minimum population which is inversely proportional to the number of characters in the search query (e.g. `t`: population > 1,000,000; `to`: population > 500,000)

**Comment by derhuerst** [comment]
@brianshaler

A _hifo_ would have to wait until the previous stream `end`s. Only then, it can forward the highest `n` items.

> My use case is a leveldb+websockets city name autocomplete feature, where it may take a couple seconds for leveldb to finish streaming all t\* cities, and I would prefer not to wait until the leveldb stream completes before I send the first result to the client.

_hifo_ can't help you with this because it is supposed to find the highest `n` of _all_ elements. Therefore, by definition, it can only "be sure" to have the highest `n` after is has "seen" _all_ elements.

**Comment by derhuerst** [comment]
Or am I missing a point here?

**Comment by brianshaler** [comment]
> Or am I missing a point here?

Not necessarily. What I'm looking for might just be a different type of thing. I think a stream interface could still be applied (where all inputs are bottled up and upon receiving an `end` event, the items in the array are emitted) but it wouldn't help much if you'll be waiting for all inputs to be read anyway.

> _hifo_ can't help you with this because it is supposed to find the highest `n` of all elements.

This only depends on if you define "highest" temporally. During a stream, which doesn't necessarily have to have an end, you can define it as the highest you have seen so far.

Maybe I'm misinterpreting _hifo_ and it only refers to the end result of a finite series. ¯_(ツ)_/¯

**Comment by derhuerst** [comment]
@brianshaler

> This only depends on if you define "highest" temporally. During a stream, which doesn't necessarily have to have an end, you can define it as the highest you have seen so far.

You can do this, but it will at least forward you every of the first elements until the array has `n` elements. From that point on, you will only get the highest. Of course, one could check if the array has `n` elements and only then forward the data, but right now this module is very tiny and supporting stream would bloat it.

What do you think about a streams wrapper called _hifo-stream_?

**Comment by brianshaler** [comment]
#### [hifo-stream](https://github.com/brianshaler/hifo-stream)

Started writing a wrapper, but I had to change how the `add` method was working (instead of returning this to be chainable, it returns an index or undefined, so the hifo-stream knows whether or not to emit an event), so it didn't end up depending on hifo.

**Comment by derhuerst** [comment]
I accepted your pull request #3 and released [version `0.4.0`](https://github.com/derhuerst/hifo/releases/tag/0.4.0).
